File: year_2022/day_14/day14.go
package main

import (
	"aoc/tools"
	"fmt"
	"regexp"
)

type cave struct {
	grid   [][]bool
	blocks int
	done   bool
}

func (c *cave) increaseDepth(rows int) {
	for i := 0; i < rows; i++ {
		newRow := make([]bool, len(c.grid[0]))
		c.grid = append(c.grid, newRow)
	}
}

func (c *cave) fall(row, column int) {
	switch {
	case c.grid[0][column] || row == len(c.grid):
		c.done = true
	case !c.grid[row][column]:
		c.fall(row+1, column)
	case !c.grid[row][column-1]:
		c.fall(row, column-1)
	case !c.grid[row][column+1]:
		c.fall(row, column+1)
	default:
		c.grid[row-1][column] = true
		c.blocks++
	}
}

func solve(input string, part2 bool) int {
	data := tools.ReadLines(input)
	nums := regexp.MustCompile(`\d+`)
	row := make([]bool, 700)
	cavern := cave{[][]bool{row}, 0, false}

	for _, line := range data {
		values := tools.StrToInt(nums.FindAllString(line, -1))

		for i := 0; i < len(values)-2; i += 2 {
			wallX, wallY := values[i], values[i+1]
			nextX, nextY := values[i+2], values[i+3]
			minX, maxX := min(wallX, nextX), max(wallX, nextX)
			minY, maxY := min(wallY, nextY), max(wallY, nextY)

			if maxY > len(cavern.grid) {
				add := maxY + 1 - len(cavern.grid)
				cavern.increaseDepth(add)
			}

			for i := minX; i <= maxX; i++ {
				cavern.grid[wallY][i] = true
			}

			for i := minY; i <= maxY; i++ {
				cavern.grid[i][wallX] = true
			}
		}
	}

	if part2 {
		cavern.increaseDepth(2)

		for i := 0; i < len(cavern.grid[0]); i++ {
			cavern.grid[len(cavern.grid)-1][i] = true
		}
	}

	for !cavern.done {
		cavern.fall(0, 500)
	}

	return cavern.blocks
}

func main() {
	testData := "test.txt"
	data := tools.GetData(2022, 14)

	fmt.Println(solve(testData, false), solve(testData, true))
	fmt.Println(solve(data, false), solve(data, true))
}

File: year_2022/day_14/day14.py
import numpy as np

from tools import loader, parsers

# np.set_printoptions(threshold=np.inf, linewidth=200)


class Cave:
    def __init__(self, data: list[str], part: int) -> None:
        """Right border is hardcoded as an arbitrarily large value,
        since adding dynamic stretching adds extra complexity to the code.
        May need to increase it if code fails."""
        self.left, self.right = 500, 700
        self.cavern = np.zeros((1, self.right), dtype=int)
        self.draw_cavern(
            [[list(map(int, coord.split(','))) for coord in line.split(' -> ')] for line in data]
        )
        if part == 2:
            self.increase_depth(1, 0)
            self.increase_depth(1, 1)

    def draw_cavern(self, data_clean: list[list[list[int]]]) -> None:
        for line in data_clean:
            for i, (x, y) in enumerate(line):
                try:
                    next_x, next_y = line[i + 1][0], line[i + 1][1]
                except IndexError:
                    break
                max_x, min_x, max_y = max(x, next_x), min(x, next_x), max(y, next_y)
                self.left = min(min_x, self.left)
                self.right = max(max_x, self.right)
                if max_y > len(self.cavern):
                    add = max_y + 1 - len(self.cavern)
                    self.increase_depth(add, 0)

                slice_x = np.s_[y, next_x:x + 1] if x - next_x > 0 else np.s_[y, x:next_x + 1]
                self.cavern[slice_x] = 1

                slice_y = np.s_[next_y:y + 1, x] if y - next_y > 0 else np.s_[y:next_y + 1, x]
                self.cavern[slice_y] = 1

    def increase_depth(self, lines: int, value: int) -> None:
        self.cavern = np.pad(self.cavern, [(0, lines), (0, 0)], constant_values=value)

    def fall(self, row: int, column: int) -> None:
        if self.cavern[0][column] == 1:
            raise IndexError('reached the top')
        if self.cavern[row][column] != 1:
            self.fall(row + 1, column)
        elif self.cavern[row][column - 1] != 1:
            self.fall(row, column - 1)
        elif self.cavern[row][column + 1] != 1:
            self.fall(row, column + 1)
        else:
            self.cavern[row - 1][column] = 1
            self.left = min(self.left, column)
            self.right = max(self.right, column)

    def go(self) -> int:
        """test part 1:
        >>> print(Cave(parsers.lines('test.txt'), part=1).go())
        24

        test part 2:
        >>> print(Cave(parsers.lines('test.txt'), part=2).go())
        93"""
        counter = 0
        while True:
            try:
                self.fall(0, 500)
                counter += 1
            except IndexError:
                break
        return counter


print(Cave(parsers.lines(loader.get()), part=1).go())  # 774
print(Cave(parsers.lines(loader.get()), part=2).go())  # 22499
